File: clojure/robot-simulator/src/robot_simulator.clj
(ns robot-simulator)

(defn robot [coordinates bearing]
  {:coordinates coordinates :bearing bearing})

(def turn-right
  {:north :east, :east :south,
   :south :west, :west :north})

(def turn-left (zipmap (vals turn-right) (keys turn-right)))

(def advance
  {:north #(update % :y inc)
   :east  #(update % :x inc)
   :south #(update % :y dec)
   :west  #(update % :x dec)})

(defn command [robot cmd]
  (case cmd
    \A (->> (:bearing robot) advance (update robot :coordinates))
    \R (->> turn-right (update robot :bearing))
    \L (->> turn-left (update robot :bearing))))

(defn simulate [commands robot]
  (->> commands (reduce command robot)))

File: r/robot-simulator/robot-simulator.R
coords = list('NORTH'=c(0,1), 'EAST'=c(1,0), 'SOUTH'=c(0,-1), 'WEST'=c(-1,0))
turn = list('NORTH'=0, 'EAST'=1, 'SOUTH'=2, 'WEST'=3)
directions = c('NORTH', 'EAST', 'SOUTH', 'WEST')

new_robot = function(coordinates, direction) structure(list(coordinates=coordinates, direction=direction), class = 'robot')

turnMod = function(direction, inc) (direction + inc) %% 4 + 1

move = function(robo, commands) {
  for (com in strsplit(commands,'')[[1]]) {
    if (com == 'A') robo$coordinates = robo$coordinates + coords[[robo$direction]]
    else if (com == 'R') robo$direction = directions[turnMod(turn[[robo$direction]], 1)]
    else robo$direction = directions[turnMod(turn[[robo$direction]], 3)]
  }
  robo
}
